File: backend/src/main/java/com/career/recommendation/service/RoadmapProgressService.java
package com.career.recommendation.service;

import com.career.recommendation.model.RoadmapProgress;
import com.career.recommendation.repository.RoadmapProgressRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import java.time.LocalDateTime;
import java.util.List;

@Service
public class RoadmapProgressService {
    @Autowired private RoadmapProgressRepository repository;
    public List<RoadmapProgress> getByStudentId(Long studentId) {
        if (studentId == null) throw new IllegalArgumentException("Student ID is required");
        return repository.findByStudentId(studentId);
    }
    public RoadmapProgress updateProgress(Long studentId, Long roadmapId, Boolean completed) {
        if (studentId == null) throw new IllegalArgumentException("Student ID is required");
        if (roadmapId == null) throw new IllegalArgumentException("Roadmap ID is required");
        boolean done = Boolean.TRUE.equals(completed);
        RoadmapProgress progress = repository.findByStudentIdAndRoadmapId(studentId, roadmapId).orElseGet(() -> {
            RoadmapProgress p = new RoadmapProgress();
            p.setStudentId(studentId); p.setRoadmapId(roadmapId); return p;
        });
        progress.setCompleted(done);
        progress.setCompletedAt(done ? LocalDateTime.now() : null);
        return repository.save(progress);
    }
}

File: frontend/src/services/roadmapProgressService.js
import api from "./api";

const roadmapProgressService = {
    // ==========================================
    // GET PROGRESS BY STUDENT
    // ==========================================

    getByStudentId: async (studentId) => {
        if (!studentId) {
            throw new Error("Student ID is required");
        }

        const response = await api.get(
            `/roadmap-progress/student/${studentId}`
        );

        return response.data;
    },

    // ==========================================
    // UPDATE PROGRESS
    // ==========================================

    updateProgress: async (
        studentId,
        roadmapId,
        completed
    ) => {
        if (!studentId) {
            throw new Error("Student ID is required");
        }

        if (!roadmapId) {
            throw new Error("Roadmap ID is required");
        }

        const response = await api.post(
            "/roadmap-progress/update",
            {
                studentId: Number(studentId),
                roadmapId: Number(roadmapId),
                completed: Boolean(completed)
            }
        );

        return response.data;
    }
};

export default roadmapProgressService;